### PR TITLE
feat: RUN-1039: Convert executed instructions to cycles differently for Wasm64

### PR DESCRIPTION
**Pull request by alexandru-uta**
In this PR the conversion from executed instructions to cycles is treated differently for Wasm64 compared to Wasm32. This PR puts in place all the scaffolding and infrastructure that allows Wasm64 to be treated differently. 

This PR inspects the Wasm on the replica side before `install_code` to determine what kind of charging it should apply.
In follow-up work we will either (1) keep this process as-is but then pass on the unzipped Wasm to the sandbox; (2) Call into the sandbox to get this information. 

The most important part of the change is related to `cycles_account_manager/lib.rs`, `install.rs`, `install_code.rs`, `update.rs`, all the other changes are mostly mechanical changes.

### CODE REVIEW

File: rs/config/src/subnet_config.rs [hunk 1]
```diff
@@ -160,6 +160,11 @@ pub const DEFAULT_UPLOAD_CHUNK_INSTRUCTIONS: NumInstructions = NumInstructions::
 pub const DEFAULT_CANISTERS_SNAPSHOT_BASELINE_INSTRUCTIONS: NumInstructions =
     NumInstructions::new(2_000_000_000);
 
+/// The cycle cost overhead of executing canister instructions when running in Wasm64 mode.
+/// This overhead is a multiplier over the cost of executing the same instructions
+/// in Wasm32 mode.
```
Comment (dsarlis) [inline rs/config/src/subnet_config.rs:165]: Should we also mention a bit why there's an overhead? Basically the bounds check that adds it.

File: rs/test_utilities/execution_environment/src/lib.rs [hunk 2]
```diff
@@ -275,6 +275,21 @@ impl ExecutionTest {
...
+            state
+                .canister_state(&canister_id)
+                .as_ref()
+                .map_or(false, |canister| {
+                    canister
+                        .execution_state
+                        .as_ref()
+                        .map_or(false, |execution_state| execution_state.is_wasm64)
+                })
+        })
```
Comment (dsarlis) [inline rs/test_utilities/execution_environment/src/lib.rs:290]: Nit: Maybe this is a case where chaining too much obscures readability. What if we break it down? Dumb code is at least easy to read.

File: rs/execution_environment/src/scheduler/test_utilities.rs [hunk 3]
```diff
@@ -197,9 +197,12 @@ impl SchedulerTest {
     }
 
     pub fn execution_cost(&self, num_instructions: NumInstructions) -> Cycles {
-        self.scheduler
-            .cycles_account_manager
-            .execution_cost(num_instructions, self.subnet_size())
+        const IS_WASM64_EXECUTION: bool = false;
```
Comment (dsarlis) [inline rs/execution_environment/src/scheduler/test_utilities.rs:200]: Nit: Is this const really necessary? We use it directly below and it's a test function.

File: rs/execution_environment/src/execution/upgrade.rs [hunk 4]
```diff
@@ -106,13 +106,15 @@ pub(crate) fn execute_upgrade(
 
     // Stage 0: validate input.
     if let Err(err) = helper.validate_input(&original, &round, round_limits) {
+        let is_wasm64_execution = original.is_wasm64_execution;
```
Comment (dsarlis) [inline rs/execution_environment/src/execution/upgrade.rs:109]: Nit: We can directly use `original.is_wasm64_execution` below and it's quite readable imo. Now there's an extra line that doesn't bring much.

File: rs/execution_environment/src/canister_manager/tests.rs [hunk 5]
```diff
@@ -7865,3 +7892,58 @@ fn chunk_store_counts_against_subnet_memory_in_initial_round_computation() {
         .unwrap_err();
     assert_eq!(error.code(), ErrorCode::SubnetOversubscribed);
 }
+
+fn run_canister_in_wasm_mode(is_wasm64_mode: bool, execute_ingress: bool) -> (Cycles, Cycles) {
```
Comment (dsarlis) [inline rs/execution_environment/src/canister_manager/tests.rs:7896]: Maybe leave a small doc comment that explains what this helper is doing?

File: rs/execution_environment/src/canister_manager.rs [hunk 6]
```diff
@@ -875,6 +878,39 @@ impl CanisterManager {
         }
     }
 
+    /// Checks if the given wasm module is a Wasm64 module.
+    /// This is solely for the purpose of install code, when at the replica level
+    /// we don't know yet if the module is Wasm32/64 and we need to prepay accordingly.
+    /// In case of errors, we simply return false, assuming Wasm32.
+    /// The errors will be caught and handled by the sandbox.
```
Comment (dsarlis) [inline rs/execution_environment/src/canister_manager.rs:885]:
```suggestion
    /// The errors will be caught and handled by the sandbox later.
```

**Review by dsarlis — COMMENTED** [review]
Would it make sense to use an enum which captures whether the canister is `Wasm32` or `Wasm64`? I feel like it could improve readability of the code a bit.

File: rs/config/src/subnet_config.rs (hunk 1 above)
Comment (alexandru-uta) [reply rs/config/src/subnet_config.rs]: makes sense, thanks. fixed now.

File: rs/test_utilities/execution_environment/src/lib.rs (hunk 2 above)
Comment (alexandru-uta) [reply rs/test_utilities/execution_environment/src/lib.rs]: thanks, I tried to make it more clear now.

File: rs/execution_environment/src/scheduler/test_utilities.rs (hunk 3 above)
Comment (alexandru-uta) [reply rs/execution_environment/src/scheduler/test_utilities.rs]: fixed.

File: rs/execution_environment/src/canister_manager/tests.rs (hunk 5 above)
Comment (alexandru-uta) [reply rs/execution_environment/src/canister_manager/tests.rs:7898]: done, thanks.

File: rs/cycles_account_manager/src/lib.rs [hunk 7]
```diff
@@ -65,6 +65,24 @@ impl std::fmt::Display for CyclesAccountManagerError {
     }
 }
 
+/// Keeps track of how a canister is executing such that the cycles account
+/// manager can charge cycles accordingly.
```
Comment (dsarlis) [inline rs/cycles_account_manager/src/lib.rs:69]:
```suggestion
/// Keeps track of how a canister is executing such that the `CyclesAccountManager`
/// can charge cycles accordingly.
```

File: rs/execution_environment/tests/subnet_size_test.rs [hunk 8]
```diff
@@ -751,6 +754,7 @@ fn get_cycles_account_manager_config(subnet_type: SubnetType) -> CyclesAccountMa
             // verified subnet is using.
             update_message_execution_fee: Cycles::new(5_000_000),
             ten_update_instructions_execution_fee: Cycles::new(10),
+            ten_update_instructions_execution_fee_wasm64: Cycles::new(20),
```
Comment (dsarlis) [inline rs/execution_environment/tests/subnet_size_test.rs:757]: Should we also do the same where we use a local variable for the old fee and set the new to be 2x that?

File: rs/execution_environment/src/execution/upgrade.rs (hunk 4 above)
Comment (dsarlis) [reply rs/execution_environment/src/execution/upgrade.rs]: But then can we not simply use `original.is_wasm64_execution` inside `finish_err`?

**Review by dsarlis — APPROVED** [review]

File: rs/execution_environment/src/execution/upgrade.rs (hunk 4 above)
Comment (alexandru-uta) [reply rs/execution_environment/src/execution/upgrade.rs]: yes, absolutely correct! fixed.

File: rs/execution_environment/tests/subnet_size_test.rs (hunk 8 above)
Comment (alexandru-uta) [reply rs/execution_environment/tests/subnet_size_test.rs]: yes, fixed, thanks.

**Pull request merged**